src/app/archive/archive_lib/processor.ml

```diff
@@ -1547,6 +1556,74 @@ module Block = struct
          |sql})
       (parent_id, State_hash.to_base58_check parent_hash)
 
+  let get_subchain (module Conn : CONNECTION) ~start_block_id ~end_block_id =
+    Conn.collect_list
+      (Caqti_request.collect Caqti_type.(tup2 int int) typ
+      {sql| WITH RECURSIVE chain AS (
+              SELECT id,state_hash,parent_id,parent_hash,creator_id,block_winner_id,snarked_ledger_hash_id,staking_epoch_data_id,
+                     next_epoch_data_id,ledger_hash,height,global_slot,global_slot_since_genesis,timestamp, chain_status
+              FROM blocks b WHERE b.id = $1
+
+              UNION ALL
+
+              SELECT b.id,b.state_hash,b.parent_id,b.parent_hash,b.creator_id,b.block_winner_id,b.snarked_ledger_hash_id,b.staking_epoch_data_id,
+                     b.next_epoch_data_id,b.ledger_hash,b.height,b.global_slot,b.global_slot_since_genesis,b.timestamp,b.chain_status
+              FROM blocks b
+
+              INNER JOIN chain
+
+              ON b.id = chain.parent_id AND (chain.id <> $2 OR b.id = $2)
+
+           )
+
+           SELECT state_hash,parent_id,parent_hash,creator_id,block_winner_id,snarked_ledger_hash_id,staking_epoch_data_id,
+                  next_epoch_data_id,ledger_hash,height,global_slot,global_slot_since_genesis,timestamp,chain_status
+           FROM chain ORDER BY height ASC
+      |sql})
+      (end_block_id,start_block_id)
+
+  let get_highest_canonical_block (module Conn : CONNECTION) =
+    Conn.find
+      (Caqti_request.find Caqti_type.unit Caqti_type.(tup2 int int64)
+         "SELECT id,height FROM blocks WHERE chain_status='canonical' ORDER BY height DESC LIMIT 1")
+
+  let mark_as_canonical (module Conn : CONNECTION) ~state_hash =
+    Conn.exec
+      (Caqti_request.exec Caqti_type.string
+         "UPDATE blocks SET chain_status='canonical' WHERE state_hash = ?")
+      state_hash
+
+  let mark_as_orphaned (module Conn : CONNECTION) ~state_hash ~height =
+    Conn.exec
+      (Caqti_request.exec Caqti_type.(tup2 string int64)
+         {sql| UPDATE blocks SET chain_status='orphaned'
+               WHERE height = $2
+               AND state_hash <> $1
+         |sql})
+      (state_hash,height)
+
+  (* update chain_status for blocks now known to be canonical or orphaned *)
+   let update_chain_status (module Conn : CONNECTION) ~block_id =
```

Comment:
Should we consider using a separate connection pool of size 1 for this update chain status procedure? Otherwise there will be a race when we receive a few blocks on new height, right?

Reply:
And then we need to somehow lock this? Is a transaction the right thing? Is there a way to do a pessimistic transaction or something?

Reply:
^^ since more than one daemon writes to the same archive database

Reply:
Are you worried that the additional work here would cause problems we haven't seen before?

There is a new RECURSIVE query, but once the db is migrated, the depth of that should be no more than about `k`.

Don't `Deferred` binds effectively serialize access to the existing pool?

Reply:
I'm worried that we could be concurrently handling more than one block at once -- but maybe this is unfounded. I'd be okay to speculatively land it and revisit if we see issues in QA just so we can get this out the door